csharp/src/Ice/Reference.cs

```diff
@@ -745,149 +868,143 @@ static void StartQueryOption(StringBuilder sb, ref bool firstOption)
+                        if (reference.Communicator.TraceLevels.Retry >= 1)
+                        {
+                            TraceRetry("retrying request because of retryable exception",
+                                       attempt,
+                                       retryPolicy,
+                                       exception);
+                        }
 
-                            observer?.Retried();
+                        if (connection != null || triedAllConnectors)
+                        {
+                            // Only count a retry attempt if the connection was established or if all the endpoints were
+                            // tried at least once. This ensures that we don't end up into an infinite loop for connection
+                            // establishment failures which don't result in endpoint exclusion.
+                            attempt++;
                         }
-                    }
 
-                    // No more retries or can't retry, throw the exception and return the remote exception
-                    if (lastException != null)
-                    {
-                        observer?.Failed(lastException.GetType().FullName ?? "System.Exception");
-                        throw ExceptionUtil.Throw(lastException);
-                    }
-                    else
-                    {
-                        observer?.Failed("System.Exception"); // TODO cleanup observer logic
-                        Debug.Assert(response != null && response.ResultType == ResultType.Failure);
-                        return response;
+                        if (retryPolicy.Retryable == Retryable.AfterDelay && retryPolicy.Delay != TimeSpan.Zero)
+                        {
+                            // The delay task can be canceled either by the user code using the provided cancellation
+                            // token or if the communicator is destroyed.
+                            using var tokenSource = CancellationTokenSource.CreateLinkedTokenSource(
+                                cancel,
+                                proxy.Communicator.CancellationToken);
+                            await Task.Delay(retryPolicy.Delay, tokenSource.Token).ConfigureAwait(false);
+                        }
+
+                        observer?.Retried();
+
+                        // TODO remove this when we implement the locator cache serial, that will retrieve a fresh endpoint
+                        // based on the previous serial.
+                        if (cached && reference.IsIndirect)
+                        {
+                            reference.LocatorInfo?.ClearCache(reference);
+                        }
+
+                        if (!reference.IsFixed && connection != null)
+                        {
+                            // Retry with a new connection!
+                            reference.ClearConnection(connection);
+                            connection = null;
```

Comment:
Do we actually sill need to clear the connection here?

Reply:
I will add a TODO, I think we do until we add PreferExistingConnection

Reply:
actually is not required anymore.